anvil/eth: Use the raw `v` signature value instead of `bool`

ngotchac opened the pull request:
Fixes https://github.com/foundry-rs/foundry/issues/7898

<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

Cf. the linked issue. The `v` value (which should be deprecated for *some* tx types) was wrongly set as the y parity byte converted into a uint.

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

Instead of converting the boolean corresponding to the y parity byte to a u64/U256, use the raw `v` value for the `v` field when available.

mattsse commented on crates/anvil/core/src/eth/transaction/mod.rs:
```diff
@@ -401,6 +401,14 @@ pub fn to_alloy_transaction_with_hash_and_sender(
     }
 }
 
+fn v_parity_to_u64(v: Parity) -> u64 {
+    match v {
+        Parity::Eip155(v) => v,
+        Parity::NonEip155(b) => (b as u64) + 27,
+        Parity::Parity(b) => b as u64,
+    }
+}
```
@prestwich is this something we should add as function instead?

prestwich replied:
yes, in [this file](https://github.com/alloy-rs/core/blob/main/crates/primitives/src/signature/parity.rs)

ngotchac replied:
Actually, to method `to_u64()` already existed ; I completely missed it :see_no_evil: Updated the PR!

ngotchac commented:
I added a 2nd commit that fixes invalid (from geth's PoV) dummy signatures. The issue was that it was using `v == 0`, while this snippet in geth'd code: https://github.com/ethereum/go-ethereum/blob/d2f00cb54edc4486314c25d9e6c5b739009c2201/core/types/transaction.go#L234-L238 was computing `plainV = byte(v.Uint64() - 27)` which made it invalid.

By using `Parity::NonEip155(false)`, `v` is actually encoded as `27` instead of `0`, which makes the validation pass.

ngotchac commented:
OK and one last commit, to actually use the provided signature when creating a transaction. This resolves an issue where the computed transaction hash in `anvil` (computed once, reused everywhere afterward) wouldn't match a computed hash from the transactions' details (eg. from a `eth_getBlock` call). In geth's Go library, a transaction hash is (almost) never parsed from the RPC response, but recomputed. This caused mismatches and all kinds of troubles.

This is a bit tangential to the initial issue, so I could if needed open another PR for it.

mattsse reviewed (approved):
lgtm

mattsse commented:
I think the failing tests are related

ngotchac commented:
> I think the failing tests are related

You mean *un*related? The only failing ones are integration tests which pass after retry.

ngotchac commented:
OK, I found the issue! So TIL, for all transactions that are not legacy, the `v` value for the signature should actually just be the parity byte (so `0` or `1`). This means that the impersonate signature cannot be constant, but depends on the transaction type. I removed some code (2nd commit) that did seem really relevant, and simplified things so that the impersonate signature is creating within the `api.rs` file, dependent on the transaction type.

To see the issue, cf. the signature change in the failing test. The previous signature wasn't actually valid. This can be checked here, for example: https://toolkit.abdk.consulting/ethereum#rlp,transaction (the v value was set to `62710` while it was an EIP-1559 tx).

zerosnacks commented:
Hi @ngotchac thanks for the PR! Would you be able to fix the small merge conflicts so we can get it reviewed and merged?

mattsse commented:
whoops, sorry I dropped this
mind bringing this over the line @zerosnacks ?

zerosnacks commented:
> whoops, sorry I dropped this mind bringing this over the line @zerosnacks ?

Done, ready for review :)
